apache/zeppelin
# [ZEPPELIN-6053] Fix(zeppelin-web): Upgrade webpack-dev-server to enable HMR

### What is this PR for?
- Fix : https://issues.apache.org/jira/browse/ZEPPELIN-6052

Background
---
Currently, the Zeppelin web application uses Webpack for its build tools. Webpack provides valuable features like hot module replacement (HMR) through webpack-dev-server, which greatly enhances the development experience by automatically updating the UI in response to code changes.

Unfortunately, hot module replacement is not functioning properly, impacting the developer experience, particularly on the frontend (zeppelin-web).

Resolution
---

To resolve this issue, I updated webpack-dev-server from version ^3.1.10 to ^4.15.2. This update was selected as the latest version compatible with Webpack version 4.

- Version Upgrade: Although I considered further upgrades, I chose the latest version compatible with Webpack version 4. I plan to migrate to Webpack version 5 incrementally.
- Dependencies: The updated version 4.15.2 requires @webpack-cli/serve version ^1.2.0, which necessitated the installation of webpack-cli@4.x.x. Consequently, I installed webpack-cli@4.10.0 too.
- Configuration Changes: Due to BREAKING CHANGES introduced by the new version, modifications were required in the webpack.config.js to accommodate these updates. I followed [this migration guide](https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md).

AS-IS
---
![화면 기록 2024-08-11 오후 3 14 14](https://github.com/user-attachments/assets/b6d646e7-0019-4f03-a714-736a403e72ea)


TO-BE
---
![화면 기록 2024-08-11 오후 1 57 45](https://github.com/user-attachments/assets/a0446451-05a9-4983-b94a-f19b8e66208b)



### What type of PR is it?
Bug Fix

### Todos
* [x] Upgrade webpack-dev-server (3.1.10 -> 4.15.2)
* [x] Install @webpack-cli/serve 
* [x] Upgrade webpack-cli (3.3.12 -> 4.10.0)

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-6053

### How should this be tested?
```
1. git clone latest code.
git clone https://github.com/apache/zeppelin.git

2. change directory to zeppelin-web
cd zeppelin-web

3. install dependencies
npm install

4. run dev mode
npm run dev

5. modify any codes you want to see the change (I'd recommend <H1/> simply)
- 1. move to zeppelin-web/src/app/home.html
- 2. change any html codes you can easily detect

6. you can check it doesn't update UI automatically. you should reload instead.
```


### Screenshots (if appropriate)

### Questions:
* Does the license files need to update?
* Is there breaking changes for older versions?
* Does this needs documentation?
